crates/testing/src/script.rs

```diff
@@ -108,20 +112,53 @@ pub async fn run_test_script<TYPES, S: TaskState<Event = Arc<HotShotEvent<TYPES>
         }
 
         for assert in &stage.outputs {
-            if let Ok(Ok(received_output)) =
-                async_timeout(Duration::from_millis(250), test_receiver.recv_direct()).await
-            {
-                tracing::debug!("Test received: {:?}", received_output);
-                validate_output_or_panic(stage_number, &received_output, assert);
-                if !task.state_mut().filter(&received_output.clone()) {
-                    tracing::debug!("Test sent: {:?}", received_output.clone());
-
-                    if let Some(res) = S::handle_event(received_output.clone(), &mut task).await {
-                        task.state_mut().handle_result(&res).await;
+            match assert {
+                EventPredicate::One(assert) => {
+                    match async_timeout(
+                        Duration::from_millis(RECV_TIMEOUT_MILLIS),
+                        test_receiver.recv_direct(),
+                    )
+                    .await
+                    {
+                        Ok(Ok(received_output)) => {
+                            tracing::debug!("Test received: {:?}", received_output);
+                            validate_output_or_panic(stage_number, &received_output, assert);
+                        }
+                        _ => panic_missing_output(stage_number, assert),
+                    }
+                }
+                EventPredicate::Consecutive(asserts) => {
+                    match async_timeout(
```

Comment:
So is this essentially waiting until the first event happens, then waiting until the second event happens? I am having trouble understanding how this differs from two events just coexisting in the `output` vector. Is this essentially *forcing* the two events to occur right after each other?

Reply:
There are two "outputs", one is the expected output and the other is the received output. For the expected output, it's just the `asserts` tuple. However, we have to call `recv_direct` twice to get the received output, which we can't just get from an `output` vector but need to construct it. The two events don't really coexist, so yes, they occur right after each other, but we just don't want to force the order.